packages/devtools_app/lib/src/memory/flutter/memory_chart.dart

```diff
@@ -0,0 +1,342 @@
+    final desc = Description()..enabled = false;
+
+    _chartController = LineChartController(
+        axisLeftSettingFunction: (axisLeft, controller) {
+          axisLeft
+            ..position = YAxisLabelPosition.OUTSIDE_CHART
+            ..setValueFormatter(LargeValueFormatter())
+            ..drawGridLines = (true)
+            ..granularityEnabled = (true)
+            ..setStartAtZero(
+                true) // Set to baseline min and auto track max axis range.
+            ..textColor = const Color.fromARGB(255, 0, 0, 0);
+        },
+        axisRightSettingFunction: (axisRight, controller) {
+          axisRight.enabled = false;
+        },
+        xAxisSettingFunction: (xAxis, controller) {
+          xAxis
+            ..position = XAxisPosition.BOTTOM
+            ..textSize = 10
+            ..textColor = ColorUtils.WHITE
+            ..drawAxisLine = false
+            ..drawGridLines = true
+            ..textColor = const Color.fromARGB(255, 0, 0, 0)
+            ..centerAxisLabels = true
+            ..setGranularity(1)
+            ..setValueFormatter(XAxisFormatter());
+        },
+        legendSettingFunction: (legend, controller) {
+          legend.enabled = false;
+          // TODO(terry): Need to support legend with a smaller text size.
+          // legend
+          //   ..shape = LegendForm.LINE
+          //   ..verticalAlignment = LegendVerticalAlignment.TOP
+          //   ..enabled = true
+          //   ..orientation = LegendOrientation.HORIZONTAL
+          //   ..typeface = legendTypeFace
+          //   ..xOffset = 20
+          //   ..drawInside = false
+          //   ..horizontalAlignment = LegendHorizontalAlignment.CENTER
+          //   ..textSize = 6.0;
+        },
+        highLightPerTapEnabled: true,
+        backgroundColor: ColorUtils.WHITE,
+        drawGridBackground: false,
+        dragXEnabled: true,
+        dragYEnabled: true,
+        scaleXEnabled: true,
+        scaleYEnabled: true,
+        pinchZoomEnabled: false,
+        description: desc);
+
+    // Compute padding around chart.
+    _chartController.setViewPortOffsets(50, 10, 10, 30);
+  }
+
+  // Datapoint entry for each used heap value.
+  final List<Entry> _used = <Entry>[];
+
+  // Datapoint entry for each capacity heap value.
+  final List<Entry> _capacity = <Entry>[];
+
+  // Datapoint entry for each external memory value.
+  final List<Entry> _externalHeap = <Entry>[];
+
+  // Trace #1 Heap Used.
+  LineDataSet usedHeapSet;
+
+  // Trace #2 Heap Capacity.
+  LineDataSet capacityHeapSet;
+
+  // Trace #3 External Memory used.
+  LineDataSet externalMemorySet;
+
+  // Index into the raw data.
+  int timerDataIndex;
+
+  // Active timer running.
+  Timer _timer;
+
+  Future<void> startFeed() async {
+    // Fetch from the beginning of the canned data for the live feed.
+    timerDataIndex = 0;
+
+    // Average rate is ~500-600 ms?
+    _timer = Timer.periodic(const Duration(milliseconds: 100), (Timer timer) {
```

Comment:
Summarizing our offline discussion on this:

The data collection for the chart should follow a notifier pattern: we want to tell the chart that new data is available and it should repaint whenever new data becomes available, rather than polling every fraction of a second.  Right now, the chart repaint events may happen with considerable lag from when new data arrives.

Can you add a TODO to switch from a Timer to a notifier?

Reply:
Yes, I also think the Timer should be part of the controller not the State.  Will add both to the TODO.